tests/pytest_tests/unit_tests/test_launch/test_controller/test_base.py

```diff
@@ -132,3 +164,168 @@ async def test_launch_item(mocked_test_manager, mocker):
     assert mock_launch_project_instance.mock_label_job.call_count == 1
     assert mocked_test_manager.legacy.runner.run.call_count == 1
     assert mocked_test_manager.active_runs.get("test-id") is not None
+
+
+@pytest.mark.parametrize(
+    "called_init_value, run_id, status, check_init_count, failed_args",
+    [
+        (True, "test-id", "finished", 1, None),
+        (
+            False,
+            "test-id",
+            "finished",
+            1,
+            (
+                "test-rqi-id",
+                "The submitted job exited successfully but failed to call wandb.init",
+                "run",
+                None,
+            ),
+        ),
+        (
+            False,
+            "test-id",
+            "stopped",
+            1,
+            (
+                "test-rqi-id",
+                "The submitted job failed to call wandb.init exited with status: stopped",
+                "run",
+                None,
+            ),
+        ),
+        (
+            False,
+            None,
+            "finished",
+            0,
+            (
+                "test-rqi-id",
+                "The submitted job was finished without assigned project or run id",
+                "agent",
+            ),
+        ),
+    ],
+)
+@pytest.mark.asyncio
+async def test_finish_launched_run(
+    mocked_test_manager,
+    mock_run_with_tracker,
+    mocker,
+    called_init_value,
+    run_id,
+    status,
+    check_init_count,
+    failed_args,
+):
+    mock_api = AsyncMock()
+    mock_fail_run_queue_item = AsyncMock()
+    mock_api.fail_run_queue_item = mock_fail_run_queue_item
+    mocked_test_manager.jobset.api = mock_api
+
+    mock_run_called_init = AsyncMock()
+    mock_run_called_init.return_value = (called_init_value, None)
+    mocker.patch(
+        "wandb.sdk.launch.agent2.controllers.base.check_run_called_init",
+        mock_run_called_init,
+    )
+
+    mock_run_with_tracker.tracker.run_id = run_id
+
+    await mocked_test_manager.finish_launched_run(mock_run_with_tracker, status)
+
+    assert mock_run_called_init.call_count == check_init_count
+    if failed_args is None:
+        assert mock_api.fail_run_queue_item.call_count == 0
+    else:
+        assert mock_api.fail_run_queue_item.call_count == 1
+        # index 0 because args
+        assert len(mock_api.fail_run_queue_item.call_args[0]) == len(failed_args)
+        for i, arg in enumerate(failed_args):
+            assert mock_api.fail_run_queue_item.call_args[0][i] == arg
```

Comment:
Could just be `assert mock_api.fail_run_queue_item.call_args[0] == failed_args` no?